### PR TITLE
Added support for mutually recursive function and predicate definitions

### PR DESCRIPTION
**Pull request by hra687261**
A proposition to fix the issue #548
I chose to add the `with` keyword as suggested by @claudemarche.
I don't see why a predicate can't be mutually recursive with a function and vice-versa, therefore I added one constructor for both.
Of course for now Dolmen doesn't support this, therefore the CI won't pass

### CODE REVIEW

**Comment by Halbaroth** [comment]
Great! It's quite strange that the feature has not been added before since everything works well with such a modification. 
Besides, some people use custom parsers. Modifying the legacy parser API could break temporary their stuff.  It would be better to only add this feature in Dolmen and tell Claude to use the Dolmen front-end now.

**Comment by Gbury** [comment]
one alternative syntax option would be to require something like `function f(x) : ret = ... and function g(y) : ... = ...`, i.e. the addition of the `function` keyword at each function definition should be enough to make the parsing non ambiguous, and not break anything that is currently accepted.

**Comment by hra687261** [comment]
How does this addition break anything? What am I missing? The `with` keyword was just added and didn't exist before and mutually recursive functions where simply not supported. This addition doesn't change anything to the definition of non-mutually-recursive functions and predicates.

Although in a purely subjective manner and from an aesthetic point of view I find the `with` keyword more "elegant" than `and function ...`/`and predicate ...`, I don't really have a strong opinion on the subject and don't mind changing it. 

The only ambiguity I see is that there is no distinction between predicates and functions that return bools, but semantically that doesn't change anything internally AFAIK.
https://github.com/OCamlPro/alt-ergo/blob/5f63a49c793e777f11d12228ed2aa3195d5fbbb5/src/lib/frontend/typechecker.ml#L2228-L2242

**Comment by Gbury** [comment]
> How does this addition break anything? What am I missing? The `with` keyword was just added and didn't exist before and mutually recursive functions where simply not supported. This addition doesn't change anything to the definition of non-mutually-recursive functions and predicates.

Adding a keyword is breaking in the sense that before, `with` was a valid name for a variable and/or a function, but once it is a keyword, it may not be used as a variable name anymore (changing the set of keywords of a language is almost always breaking).

**Comment by hra687261** [comment]
True, I'll change it.

**Comment by Halbaroth** [comment]
I prefer the Guillaume's syntax but I haven't a strong opinion too. In my opinion, users expect some changes when we will deprecate the legacy parser but we should'nt break the legacy parser before this change ;)

**Comment by Halbaroth** [comment]
If your PR is ready for reviewing, let me know and I'll review it right now ;)

File: src/lib/frontend/typechecker.ml [hunk 1]
```diff
@@ -2154,6 +2156,51 @@ let type_user_defined_type_body ~is_recursive env acc (loc, ls, s, body) =
     assert (not is_recursive); (* Abstract types are not recursive *)
     acc, env
 
+let type_fun_aux_1 env loc n ?ret_ty l  =
```
Comment (Gbury) [inline src/lib/frontend/typechecker.ml]: is it possible to find a better name for this function (and for `type_fun_aux_2`) ? something that would better describe what the function is doing...
Something along the lines of `declare` for `aux_1` and define for `aux_2` if I understand the code correctly.

File: src/parsers/native_parser.mly [hunk 2]
```diff
@@ -229,11 +238,21 @@ algebraic_args:
...
-and_recursive_opt:
+and_recursive_ty_opt:
   | { [] }
   | AND ty_vars = type_vars ty = ident EQUAL enum = list1_constructors_sep_bar
-others = and_recursive_opt
-    { (($startpos, $endpos), ty_vars, ty, enum) :: others}
+      others = and_recursive_ty_opt
+      { (($startpos, $endpos), ty_vars, ty, enum) :: others}
+
+with_recursive_def_opt:
```
Comment (Gbury) [inline src/parsers/native_parser.mly]:
```suggestion
and_recursive_def_opt:
```

**Comment by hra687261** [comment]
@Halbaroth it is!

File: src/lib/frontend/typechecker.ml (hunk 1 above)
Comment (hra687261) [reply src/lib/frontend/typechecker.ml]: Done.

File: src/parsers/native_parser.mly (hunk 2 above)
Comment (hra687261) [reply src/parsers/native_parser.mly]: Done.

**Review by Halbaroth — APPROVED** [review]

**Pull request merged**